File: sample-monorepo/python-fastapi/services/payments/src/refunds.py
"""POST /refunds — the refund decision lives here and nowhere else.

Decision matrix (from ``specs/feature-refunds.md``):

| Payment lookup | Payment status  | Requested amount | Decision |
|----------------|-----------------|------------------|----------|
| not found      | —               | —                | rejected |
| found          | not `completed` | —                | rejected |
| found          | `completed`     | > payment amount | rejected |
| found          | `completed`     | <= payment amount| approved |

Responses conform to ``RefundResult`` in ``contracts/refund.schema.json``.
Approved refunds also emit a ``NotificationEvent`` (see
``contracts/notification.schema.json``) to the notifications service; a
notification failure never fails the refund itself.
"""

import logging
import os
import uuid
from datetime import datetime, timezone
from typing import Optional

import httpx
from fastapi import APIRouter
from pydantic import BaseModel, Field

from src.store import get_payment

logger = logging.getLogger("payments.refunds")

router = APIRouter()

NOTIFICATIONS_URL = os.environ.get("NOTIFICATIONS_URL", "http://localhost:8002")
NOTIFICATION_TIMEOUT_SECONDS = 2.0


class RefundRequest(BaseModel):
    """Mirror of ``RefundRequest`` in contracts/refund.schema.json."""

    payment_id: str
    amount: float = Field(ge=0)
    reason: str


def decide(payment: Optional[dict], requested_amount: float) -> tuple[str, Optional[str]]:
    """Apply the decision matrix; return (status, machine-readable rejection reason)."""
    if payment is None:
        return "rejected", "payment_not_found"
    if payment["status"] != "completed":
        return "rejected", "payment_not_completed"
    if requested_amount > payment["amount"]:
        return "rejected", "amount_exceeds_payment"
    return "approved", None


def notify(event: dict) -> None:
    """Best-effort POST of a NotificationEvent to the notifications service.

    The refund is already decided when this runs, so delivery problems are
    logged and swallowed — a down notifications service must never turn an
    approved refund into an error.
    """
    try:
        response = httpx.post(
            f"{NOTIFICATIONS_URL}/events",
            json=event,
            timeout=NOTIFICATION_TIMEOUT_SECONDS,
        )
        response.raise_for_status()
    except httpx.HTTPError as exc:
        logger.warning(
            "notification for refund %s not delivered: %s", event["refund_id"], exc
        )


@router.post("/refunds")
def create_refund(request: RefundRequest) -> dict:
    """Decide the refund and return a RefundResult."""
    payment = get_payment(request.payment_id)
    status, rejection_reason = decide(payment, request.amount)

    result = {
        "refund_id": "ref-" + uuid.uuid4().hex[:8],
        "payment_id": request.payment_id,
        "status": status,
        "processed_at": datetime.now(timezone.utc)
        .isoformat(timespec="seconds")
        .replace("+00:00", "Z"),
    }
    if rejection_reason is not None:
        result["reason"] = rejection_reason

    if status == "approved":
        notify(
            {
                "event_type": "refund.approved",
                "payment_id": payment["id"],
                "refund_id": result["refund_id"],
                "recipient": payment["customer_email"],
            }
        )

    return result

File: sample-monorepo/node-express/services/payments/src/refunds.js
"use strict";

const crypto = require("node:crypto");
const express = require("express");

// Decision logic per specs/feature-refunds.md:
// approve iff the payment exists AND status === "completed" AND the requested
// amount is <= the payment amount; reject in every other case with a
// machine-readable reason.
function decide(payment, amount) {
  if (!payment) {
    return { status: "rejected", reason: "payment_not_found" };
  }
  if (payment.status !== "completed") {
    return { status: "rejected", reason: "payment_not_completed" };
  }
  if (amount > payment.amount) {
    return { status: "rejected", reason: "amount_exceeds_payment" };
  }
  return { status: "approved" };
}

// Fire-and-forget delivery: a notifications outage must never fail a refund,
// so every failure path here only logs.
async function postNotification(notificationsUrl, event) {
  try {
    const response = await fetch(`${notificationsUrl}/events`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(event),
    });
    if (!response.ok) {
      console.error(`notification rejected: HTTP ${response.status}`);
    }
  } catch (error) {
    console.error(`notification delivery failed: ${error.message}`);
  }
}

function createRefundsRouter({ store, notificationsUrl }) {
  const router = express.Router();

  router.post("/refunds", async (req, res) => {
    const { payment_id: paymentId, amount, reason } = req.body ?? {};
    if (
      typeof paymentId !== "string" ||
      typeof amount !== "number" ||
      Number.isNaN(amount) ||
      amount < 0 ||
      typeof reason !== "string"
    ) {
      res.status(422).json({ error: "body is not a valid RefundRequest" });
      return;
    }

    const payment = store.get(paymentId);
    const decision = decide(payment, amount);

    // Shape per contracts/refund.schema.json (RefundResult).
    const result = {
      refund_id: `ref-${crypto.randomUUID().slice(0, 8)}`,
      payment_id: paymentId,
      status: decision.status,
      processed_at: new Date().toISOString(),
    };
    if (decision.status === "rejected") {
      result.reason = decision.reason;
    }

    if (decision.status === "approved") {
      await postNotification(notificationsUrl, {
        event_type: "refund.approved",
        payment_id: paymentId,
        refund_id: result.refund_id,
        recipient: payment.customer_email,
      });
    }

    res.status(200).json(result);
  });

  return router;
}

module.exports = { createRefundsRouter, decide };
